File: legacy_code/diagnose_bottom_fluxes_GoT_bathy_V2.m
%%%% diagnose bottom fluxes (no point sources) %%%%%
opengl neverselect
close all
clear all
curdir=pwd;

nx=450;
ny=300;
% nz=50;
nt=365;

% Global Area (Integrated horizontal Area [m^2])
% GA=1.179E+09

% Mean (±standard deviation) annual benthic fluxes of N, P, Si and O2 (mmol m−2 d−1).
% Reference                 NO−3		NH+4		PO3−4		Si(OH)4		O2
% Bertuzzi et al. (1997)	0.17±0.73	0.8±0.7		0.029±0.05	2.59±2.3	−20.4±8.9
% BFM–POM 1D				0.27±0.16	0.63±0.38	0.048±0.03	0.47±0.41	−5.14±3.5

namef   =['N';'P';'B';'C';'O'];
patom   =[ 14; 31;  1;  1; 16];

load('mask_surface_GoT_iNEST_V2'); % surf
surf(surf==0)=NaN;

cd /share/data/squerin/esperimenti/GoT_iNEST/bathymetry

fidx=fopen('./bathydata_GoT_iNEST_V2/DXG.data','r','l');
dxg=fread(fidx,[nx,ny],'float32');
fclose(fidx);

fidy=fopen('./bathydata_GoT_iNEST_V2/DYG.data','r','l');
dyg=fread(fidy,[nx,ny],'float32');
fclose(fidy);

cd(curdir)

%%% FLUSSI %%%

for var=1:size(namef,1)
    if strcmp(namef(var),'N')
        eval(['fidin_',namef(var),...
            '=fopen(''',namef(var),'_bottom_fluxes_GoT_iNEST_V2_x2.dat'',''r'',''l'');']);
    elseif strcmp(namef(var),'P')
        eval(['fidin_',namef(var),...
            '=fopen(''',namef(var),'_bottom_fluxes_GoT_iNEST_V2_x3.dat'',''r'',''l'');']);
    else
        eval(['fidin_',namef(var),...
            '=fopen(''',namef(var),'_bottom_fluxes_GoT_iNEST_V2_flat.dat'',''r'',''l'');']);
    end
    
    tbflux=zeros(nx,ny,nt);
    tbf=zeros(nx,ny,nt);
    for t=1:nt
        eval(['bflux=fread(fidin_',namef(var),',[nx,ny],''float32'');']); % [mmol/m^2/s]
        tbflux(:,:,t)=bflux; % [mmol/m^2/s]
        tbf(:,:,t)=bflux.*dxg.*dyg; % [mmol/s]
    end
    tbfint=squeeze(sum(sum(tbf,1),2));
    
    tonnyr=sum(tbfint).*86400./1e+9.*patom(var); % [tonn/yr]
    disp([namef(var),': ',num2str(tonnyr,'%e'),' [tonn/yr]']);
    
    figure;plot(tbfint); % [mmol/s]
    title(['instant flux of ',namef(var),' - spatially integrated [mmol/s]'])
    set(gcf,'renderer','zbuffer','PaperPositionMode','auto')
    eval(['print -dpng -r600 instant_flux_',namef(var),'.png'])
    
    mbflux=mean(tbflux,3);
    figure;pcolor((mbflux.*surf)');shading flat;colorbar
    title(['mean flux of ',namef(var),' [mmol/m^2/s]'])
    set(gcf,'renderer','zbuffer','PaperPositionMode','auto')
    eval(['print -dpng -r600 mean_flux_',namef(var),'.png'])
    % da fare eventualmente plot mensili...
    
    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    
    figure;
    for stat=20:20:240
        plot(squeeze(tbflux(200,stat,:)),'b')
        hold on
        plot(repmat(squeeze(mean(tbflux(165,stat,:),3)),365),'--r')
    end
    title(['mean bottom flux of ',namef(var),' = ',num2str(mean(mean(mbflux))),' [mmol/m^2/s]'])
    set(gcf,'renderer','zbuffer','PaperPositionMode','auto')
    eval(['print -dpng -r600 time_series_flux_transect_',namef(var),'.png'])
end

File: legacy_code/python_translation/diagnose_bottom_fluxes_GoT_bathy_V2.py
# from legacy code diagnose_bottom_fluxes_GoT_bathy_V2.py
import os

import matplotlib.pyplot as plt
import numpy as np

import legacy_code.python_translation.parameters as p
import legacy_code.python_translation.pathes_to_data as ptd

# diagnose bottom fluxes (no point sources)
print("Start diagnose_bottom_fluxes_GoT_bathy_V2")
if not p.is_plot:
    print("No figures printed")
curdir = os.getcwd()

# Load surface mask
surf = np.fromfile(
    ptd.mask_path + "mask_surface_GoT_iNEST_V2", dtype="float32"
).reshape(p.ny, p.nx)
surf[surf == 0] = np.nan

# Load bathymetry data
os.chdir(ptd.bathymetry_path)

dxg = np.fromfile("DXG.data", dtype="float32").reshape(p.ny, p.nx)
dyg = np.fromfile("DYG.data", dtype="float32").reshape(p.ny, p.nx)

# FLUXES
os.chdir(ptd.bottom_fluxes_path)
for var in range(len(p.namef)):
    if p.namef[var] == "N":
        fidin = open(f"{p.namef[var]}_bottom_fluxes_GoT_iNEST_V2_x2.dat", "rb")
    elif p.namef[var] == "P":
        fidin = open(
            f"{p.namef[var]}_bottom_fluxes_GoT_iNEST_V2_x2.dat", "rb"
        )  ## io ho solo x2
    else:
        fidin = open(
            f"{p.namef[var]}_bottom_fluxes_GoT_iNEST_V2_flat.dat", "rb"
        )

    tbflux = np.zeros((p.ny, p.nx, p.nt), dtype="float32")
    tbf = np.zeros((p.ny, p.nx, p.nt), dtype="float32")
    for t in range(p.nt):
        bflux = np.fromfile(fidin, dtype="float32", count=p.nx * p.ny).reshape(
            (p.ny, p.nx)
        )  # [mmol/m^2/s]
        tbflux[:, :, t] = bflux  # [mmol/m^2/s]
        tbf[:, :, t] = bflux * dxg * dyg  # [mmol/s]
    tbfint = np.sum(tbf, axis=(0, 1))

    tonnyr = np.sum(tbfint) * 86400 / 1e9 * p.patom[var]  # [tonn/yr]
    print(f"{p.namef[var]}: {tonnyr:.2e} [tonn/yr]")

    # plt.figure()
    if p.is_plot:
        plt.plot(tbfint)  # [mmol/s]
        plt.title(
            f"instant flux of {p.namef[var]} - spatially integrated [mmol/s]"
        )
        plt.savefig(f"instant_flux_{p.namef[var]}.png", dpi=600)

    mbflux = np.mean(tbflux, axis=2)
    if p.is_plot:
        plt.figure()
        plt.pcolor(mbflux * surf)
        plt.colorbar()
        plt.title(f"mean flux of {p.namef[var]} [mmol/m^2/s]")
        plt.show()
        plt.savefig(f"mean_flux_{p.namef[var]}.png", dpi=600)

    # plt.figure()
    if p.is_plot:
        for stat in range(20, 241, 20):
            plt.plot(tbflux[200, stat, :], "b")
            plt.plot(
                np.repeat(np.mean(tbflux[165, stat, :], axis=0), 365), "--r"
            )
        plt.title(
            f"mean bottom flux of {p.namef[var]} = {np.mean(mbflux):.2f} [mmol/m^2/s]"
        )
        plt.show()
        plt.savefig(f"time_series_flux_transect_{p.namef[var]}.png", dpi=600)

    fidin.close()
print("Done diagnose_bottom_fluxes_GoT_bathy_V2")
